### PR TITLE
[WIP] Village API

### PR DESCRIPTION
**Pull request by egg82**
This adds a Bukkit (Paper) API to get and modify info on villages and sieges in the world.
There's a fair number of changes here because this update completely changes how villages work.

This should encompass everything related to villages including sieges and raids. It should also include events where applicable.

- [x] Add sieges
- [ ] Add Raids
- [ ] Add general village info

### CODE REVIEW

File: Spigot-Server-Patches/0410-Add-village-API.patch [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+-import net.minecraft.server.Ticket;
+-import net.minecraft.server.TicketType;
+-import net.minecraft.server.Unit;
+-import net.minecraft.server.Vec3D;
+-import net.minecraft.server.WorldGenFeatureEmptyConfiguration;
+-import net.minecraft.server.WorldGenHugeMushroomConfiguration;
+-import net.minecraft.server.WorldGenerator;
+-import net.minecraft.server.WorldNBTStorage;
+-import net.minecraft.server.WorldServer;
++import net.minecraft.server.*;
```
Comment (kickash32) [inline Spigot-Server-Patches/0410-Add-village-API.patch:320]: revert these import changes

File: Spigot-Server-Patches/0410-Add-village-API.patch [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
++
+                     return true;
+                 }
+             }
+@@ -127,7 +155,7 @@ public class VillageSiege {
+         return null;
+     }
+ 
+-    static enum State {
++    public static enum State { // Paper - public access
```
Comment (kickash32) [inline Spigot-Server-Patches/0410-Add-village-API.patch]: note what access type it was previously

File: Spigot-Server-Patches/0410-Add-village-API.patch [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+-import net.minecraft.server.WorldServer;
++import net.minecraft.server.*;
+ import org.apache.commons.lang.Validate;
+ import org.bukkit.BlockChangeDelegate;
+ import org.bukkit.Bukkit;
+@@ -118,6 +52,7 @@ import org.bukkit.craftbukkit.block.data.CraftBlockData;
+ import org.bukkit.craftbukkit.entity.CraftItem;
+ import org.bukkit.craftbukkit.entity.CraftLightningStrike;
+ import org.bukkit.craftbukkit.entity.CraftPlayer;
++import org.bukkit.craftbukkit.generator.CustomChunkGenerator;
```
Comment (kickash32) [inline Spigot-Server-Patches/0410-Add-village-API.patch]: missing paper comment

**Review by kickash32 — CHANGES_REQUESTED** [review]
I'm not sure how sieges work, but can there only be one siege happening in the world at a time?

**Comment by kickash32** [comment]
Related #2095 and #1491

**Comment by jacobsandersen** [comment]
Seems okay at first read through. I would +1 @kickash32's comment re: can only 1 siege per world be active (at once?)

Also, the siege provider seems to have state that is always attached to the world even if no such siege is even occurring. I question what use there would be to have that in the World class (in the API)?

**Comment by kickash32** [comment]
Also might be relevant #2230

File: Spigot-Server-Patches/0410-Add-village-API.patch (hunk 1 above)
Comment (egg82) [reply Spigot-Server-Patches/0410-Add-village-API.patch:320]: Yeah, I keep forgetting to set IJ to not do that. I'll revert once I'm done.

**Comment by egg82** [comment]
You are correct, there can only be one siege happening in the world at a time. I was just as surprised to figure that out, but the code and wikis don't lie here. It's also correct that the siege class is always instantiated - it simply gets re-used per siege. A possible optimization is not to re-wrap the VillageSiege class per call, but rather to have the CraftSiege as part of the VillageSiege class and return that instead.

The reason for having it in the World class is because it's a per-world thing. It makes sense on both an intuitive and programmatic level.

**Comment by jacobsandersen** [comment]
Right, but does it need to be in the API? I understand having it in `CraftWorld`. However, accessing it directly from `World` (again, in the API) would not make sense as you wouldn't be sure a siege is occurring or not. 

Sure, you can `setState(SiegeState)`, but does calling the setter with `ACTIVATED` /actually/ create and activate a new siege in the world?

If so, then I can see the benefit in terms of the API here. Otherwise, it seems like an implementation detail and should be left in `CraftWorld`.

To be clear, the only change I'm actively suggesting is the removal of `getSiegeProvider` from `World` (API only).

**Comment by egg82** [comment]
Yes, part of the patch is to be able to activate a siege by either setting the SiegeState to ACTIVATED or by calling tryForceSiege (both eventually call the same method to start the siege and fire the event)

**Comment by Misio12320** [comment]
Are you doing it?

**Comment by egg82** [comment]
> Are you doing it?

I haven't had the time recently, sadly. I'll hop back on this for 1.15 when I can.

**Comment by Misio12320** [comment]
Now? Or the issue is closed this mean is done and able to 1.15.2

**Comment by Misio12320** [comment]
Still in progress? :thinking:, well add also "on raid win|lose:" "on raid attack" etc. events

**Pull request closed without merging**